File: Source/Hafner.Compatibility.MetaPackage.CompileTests.VB/Hafner.Compatibility.SerializableAttribute/Struct1.vb
'GuidAttribute on struct level

Imports System
Imports System.Runtime.InteropServices

Namespace SerializableAttributeTests

    <Serializable>
    Public Structure Struct1
        Implements IEquatable(Of Struct1)

        Public Property Value As Integer

        Public Overloads Overrides Function Equals(ByVal obj As Object) As Boolean
            If (TypeOf obj Is Struct1) Then
                Dim other As Struct1 = DirectCast(obj, Struct1)
                Return Equals(other)
            End If
            Return False
        End Function

        Public Overloads Function Equals(ByVal other As Struct1) As Boolean Implements IEquatable(Of Struct1).Equals
            Return Value = other.Value
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return Value
        End Function

        Public Shared Operator =(ByVal left As Struct1, ByVal right As Struct1) As Boolean
            Return left.Equals(right)
        End Operator

        Public Shared Operator <>(ByVal left As Struct1, ByVal right As Struct1) As Boolean
            Return Not left.Equals(right)
        End Operator

    End Structure

End Namespace

File: Source/Hafner.Compatibility.MetaPackage.CompileTests.CS/Hafner.Compatibility.SerializableAttribute/Struct1.cs
//GuidAttribute on struct level

namespace Hafner.Compatibility.CompileTest.CS.SerializableAttribute;

using System;
using System.Runtime.InteropServices;

[Serializable]
public struct Struct1 : System.IEquatable<Struct1> {

    public int Value { get; set; }

    public override bool Equals(object? obj) {
        if (obj is not Struct1 other) return false;
        return Equals(other);
    }

    public bool Equals(Struct1 other) {
        return Value == other.Value;
    }

    public override int GetHashCode() {
        return Value;
    }

    public static bool operator ==(Struct1 left, Struct1 right) {
        return left.Equals(right);
    }

    public static bool operator !=(Struct1 left, Struct1 right) {
        return !left.Equals(right);
    }

}
